fix(docs-infra): more style fixes for angular.io

gkalpak opened the pull request:
Several more style fixes for angular.io:

<details>
  <summary><b>fix(docs-infra): show external link icons for external links on error pages</b> <i>(click to expand)</i></summary><br />

  This commit adds error pages to the list of docs pages that have an external link icon next to links to external URLs.

</details>

---
<details>
  <summary><b>fix(docs-infra): use relative URLs for internal links on error pages</b> <i>(click to expand)</i></summary><br />

  Previously, some of the links on the error pages had URLs prefixed with `https://angular.io/`. This caused them to be treated as external URLs, which had the following downsides:
  - Links would always point to `angular.io` instead of the same version as the error page (e.g. `next.angular.io` or `v11.angular.io`).
  - Dgeni would not be able to check that the URLs are valid (i.e. point to existing pages).
  - An external link icon would incorrectly be shown next to the links on `angular.io`.

  This commit fixes the links to use relative URLs.

</details>

---
<details>
  <summary><b>fix(docs-infra): show the "Errors List" page under "ERRORS" in search results</b> <i>(click to expand)</i></summary><br />

  Previously, the "Errors List" page was appearing under the "OTHER" section in search results.

  This commit fixes this to make it appear under the "ERRORS" section.

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107691151-c1b8ed80-6cb3-11eb-8079-fcace685f0ec.png" alt="search-results before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107691145-bfef2a00-6cb3-11eb-8523-b2747fa40469.png" alt="search-results after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>fix(docs-infra): merge `docs` with `guide` and `start` with `tutorial` in search-results</b> <i>(click to expand)</i></summary><br />

  Previously, the `docs.md` guide was appearing under "OTHER" in search results and the results for the `/start*` tutorial pages were appearing under "START".

  This commit changes this so that `docs.md` appears under "GUIDES" and `/start*` appear under "TUTORIALS", since that is where they belong conceptually.

  It also changes the header of the guides search-area from "GUIDE" to "GUIDES" and that of tutorials from "TUTORIAL" to "TUTORIALS".

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107811568-0ce80480-6d77-11eb-8652-e7a947c36e63.png" alt="search-results areas before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107811569-0eb1c800-6d77-11eb-9a69-0000a3703c8a.png" alt="search-results areas after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>refactor(docs-infra): remove unused CSS for the version selector</b> <i>(click to expand)</i></summary><br />

  This commit removes some CSS rules targeting `.doc-version select` in the sidenav. These rules do not match any elements any more, since now we use a custom `<aio-select>` component (instead of the `<select>` element).

</details>

---
<details>
  <summary><b>refactor(docs-infra): remove unnecessary `font-family` styles</b> <i>(click to expand)</i></summary><br />

  This commit removes some unnecessary styles setting `font-family` to `$main-font`. These styles are redundant, because the targeted elements already inherit this style from `<body>`.

</details>

---
<details>
  <summary><b>fix(docs-infra): ensure the main font is applied to search inputs</b> <i>(click to expand)</i></summary><br />

  It turns out that `<input>` and `<button>` elements do not inherit the `font-family` style from `<body>` by default, but rather use a user-agent defined style. This means that their font-family might be different than the one used in the rest of the page.

  This commit fixes it by ensuring `<input>` (and other) elements inherit their `font-family` style from their parent element.

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107853245-76bae980-6e1d-11eb-8318-e5f6e13876cc.png" alt="inputs font before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107853246-77538000-6e1d-11eb-86f2-e3e7e41158f5.png" alt="inputs font after" height="250" />
  </p>

  (The difference in font is subtle, but it's there.)

</details>

---
<details>
  <summary><b>refactor(docs-infra): create mixin for styling docs pages</b> <i>(click to expand)</i></summary><br />

  Previously, in order to apply some styles to docs (i.e. non-marketing) pages, we listed the various `.folder-*` classes that corresponded to docs pages. This meant that adding/removing a docs area required updates in several places, which is error-prone.

  This commit avoids this by using a Sass mixin for applying styles to docs pages.

</details>

---
<details>
  <summary><b>fix(docs-infra): add spacing between past and future events sections</b> <i>(click to expand)</i></summary><br />

  This commit adds some more spacing between the past and future events sections on the "Events" page.

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107989711-82eca580-6fdb-11eb-837e-1255d439d51a.png" alt="events spacing before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107989708-81bb7880-6fdb-11eb-8cf4-4653254e1c37.png" alt="events spacing after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>fix(docs-infra): make the "Contributors" page header similar to other marketing pages</b> <i>(click to expand)</i></summary><br />

  _**Before (wide screen):**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107983880-634f8000-6fcf-11eb-8ad9-5a5df65d3d5e.png" alt="contributors (wide screen) before" height="250" />
  </p>

  _**Before (narrow screen):**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107983893-6a768e00-6fcf-11eb-9ccd-158ec491404f.png" alt="contributors (narrow screen) before" height="250" />
  </p>

  _**After (wide screen):**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107983903-6f3b4200-6fcf-11eb-94e1-182893b7a715.png" alt="contributors (wide screen) after" height="250" />
  </p>

  _**After (narrow screen):**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107983895-6ba7bb00-6fcf-11eb-9ff5-59d221ba906d.png" alt="contributors (narrow screen) after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>fix(docs-infra): make top-nav consistent across all marketing pages</b> <i>(click to expand)</i></summary><br />

  Previously, some of the marketing pages had different styles for the top-nav than others (even if they had the same layout and blue-background header). More specifically, the top-nav had a box-shadow and it was absolutely positioned on some marketing pages, while it had no box-shadow and was statically positioned on others.

  This commit makes the appearance of marketing pages wrt the top-nav consistent across all marketing pages by changing the styles for the remaining pages:

  - Contributors (`/about`)
  - Contribute (`/contribute`)
  - Press kit (`/presskit`)

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107984898-a6aaee00-6fd1-11eb-8bf3-79393c8983ff.png" alt="contribute topnav shadow before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107984900-a7438480-6fd1-11eb-8d9b-a643d69ab692.png" alt="contribute topnav shadow after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>fix(docs-infra): use consistent padding in marketing pages</b> <i>(click to expand)</i></summary><br />

  This commit makes the padding of the "Events", "Features" and "Press kit" pages consistent with that of other marketing pages.

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107986067-f8547800-6fd3-11eb-9164-51793e431d05.png" alt="presskit padding before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107986064-f7234b00-6fd3-11eb-9050-ed7ec04e7443.png" alt="presskit padding after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>fix(docs-infra): fix top margin of "Press kit" page on small screens</b> <i>(click to expand)</i></summary><br />

  Previously, the "Press kit" page had a larger top margin on smaller screens, that seemed unnecessary.

  This commit removes the extra top margin.

  _**Before:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107988545-efb27080-6fd8-11eb-91d6-79651f2dffaf.png" alt="presskit top section margin before" height="250" />
  </p>

  _**After:**_
  <p align="center">
    <img src="https://user-images.githubusercontent.com/8604205/107988547-f04b0700-6fd8-11eb-9c4b-d444b39a34fe.png" alt="presskit top section margin after" height="250" />
  </p>

</details>

---
<details>
  <summary><b>refactor(docs-infra): remove duplicate or unused CSS styles</b> <i>(click to expand)</i></summary><br />

  This commit removes some CSS styles that have no effect (i.e. are duplicates or overridden by other rules).

</details>

---
<details>
  <summary><b>refactor(docs-infra): create mixin for styling marketing pages</b> <i>(click to expand)</i></summary><br />

  Previously, in order to apply some styles to marketing (i.e. non-docs) pages, we listed the various `.page-*` classes that corresponded to docs pages. This meant that adding/removing a marketing page required updates in several places, which is error-prone.

  This commit avoids this by using a Sass mixin for applying styles to marketing pages.

</details>

mary-poppins commented:
You can preview 4c1e0b1 at https://pr40881-4c1e0b1.ngbuilds.io/.

mary-poppins commented:
You can preview 055ac13 at https://pr40881-055ac13.ngbuilds.io/.

kapunahelewong reviewed (approved):
This is super, @gkalpak! Thank you so much! 🌼

LGTM

Reviewed-for: global-docs-approvers

petebacondarwin commented on aio/src/app/shared/search-results/search-results.component.ts:
```diff
@@ -28,9 +28,15 @@ export class SearchResultsComponent implements OnChanges {
   searchState: SearchState = SearchState.InProgress;
-  readonly topLevelFolders = ['api', 'cli', 'errors', 'guide', 'start', 'tutorial'];
+  readonly defaultFolder = 'other';
+  readonly topLevelFolders = ['api', 'cli', 'docs', 'errors', 'guide', 'start', 'tutorial'];
+  readonly folderToAreaOverrides: Record<string, string> = {
+    docs: 'guides',
+    guide: 'guides',
+    start: 'tutorials',
+    tutorial: 'tutorials',
+  };
```
How about just a single mapping:

```ts
readonly folderToAreaMap: Record<string, string> = {
    api: 'api',
    cli: 'cli',
    docs: 'guides',
    errors: 'errors'.
    guide: 'guides',
    start: 'tutorials',
    tutorial: 'tutorials',
}
```

Then you can change the code in `computeAreaName()` to something like:

```ts
  private computeAreaName(result: SearchResult): string {
    const [firstSegment, rest] = result.path.split('/', 2);
    return this.folderToAreaMap[firstSegment] || (rest && firstSegment) || this.defaultFolder;
  }
```

gkalpak replied:
Good call :+1:

`computeAreaName()` ended up as:

```ts
private computeAreaName(result: SearchResult): string {
  const [folder] = result.path.split('/', 1);
  return this.folderToAreaMap[folder] ?? this.defaultArea;
}
```

petebacondarwin commented on aio/src/app/custom-elements/events/events.component.html:
```diff
@@ -27,6 +27,8 @@ <h2>Where we'll be presenting:</h2>
   </tbody>
 </table>
 
+<br />
```
Any reason not to do this with styling rather than adding a new element?

gkalpak replied:
I was on the fence on this one too. Since this was a one-off thing, I felt it was better to make this change more in one place (in the HTML) vs adding a class in the HTML and the burring what the class does somewhere in the CSS (esp. since we don't use component-level CSS).

I don't feel strongly about this though. Happy to use a different approach if you think it is better.

petebacondarwin replied:
Well given that this is inside a template then you could add the style in quite an encapsulated way here:

https://github.com/angular/angular/blob/4f8a0abb2bb44fc1442e646e5257dc6cf98227f8/aio/src/app/custom-elements/events/events.component.ts#L36-L39

Something like?

```ts
@Component({
  selector: 'aio-events',
  templateUrl: 'events.component.html',
  styles: ['h2 ~ h2 { margin-top: 50px; }']
})
export class EventsComponent implements OnInit {
```

gkalpak replied:
I don't think this is a good idea tbh. For better or worse, we have chosen to not use component-level styles for the time being, so we should at least be consistent about that :grin:

petebacondarwin replied:
Yeah - I knew you would say that, but I couldn't help myself.
I don't feel that strongly about this here either. Happy to leave it as is.

petebacondarwin commented on aio/src/styles/1-layouts/_top-menu.scss:
```diff
@@ -26,10 +26,13 @@ mat-toolbar.app-toolbar {
+  .page-features &,
+  .page-home &,
+  .page-presskit &,
+  .page-resources & {
```
Is it worth putting these into a mixin too?

petebacondarwin replied:
LOL - I saw that you did exactly this later on.

petebacondarwin reviewed (approved):
LGTM! A couple of minor NITs that you can consider.

In this commit:

**fix(docs-infra): ensure the main font is applied to search inputs**

Given that you are also fixing "other inputs" as well, perhaps remove the word `search` from the commit title?

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

mary-poppins commented:
You can preview 4f8a0ab at https://pr40881-4f8a0ab.ngbuilds.io/.

jelbourn commented on aio/src/styles/_mixins.scss:
```diff
@@ -125,3 +125,119 @@
     }
   }
 }
+
```
I see that this file already existed, but this might be a good opportunity to move content into more specifically-named files. I consider having files named e.g. "mixins.scss" to be an anti-pattern and would recommend separate files named based on what they target. 

(could also be a follow-up at some point in the future)

gkalpak replied:
I'll leave it for a future clean-up :grin:
(TBH, I am not sure what files the mixins would be split out to 😅)

jelbourn commented on aio/src/styles/_mixins.scss:
```diff
@@ -125,3 +125,119 @@
+ *
+ * @param {boolean} nestParentSelector - If true, the parent selector (&) is nested inside the docs pages selectors.
+ */
+@mixin docsPages($nestParentSelector: false) {
```
Sass mixins are conventionally named with dash-case (though I see this file already was using camelCase for one)

gkalpak replied:
Oh, I had no idea :grin:
I updated the newly added mixins and added one more commit to change `deployTheme()` to `deploy-theme()`.

jelbourn commented on aio/src/styles/_mixins.scss:
````diff
@@ -125,3 +125,119 @@
+ * ```
+ *
+ * @param {boolean} nestParentSelector - If true, the parent selector (&) is nested inside the docs pages selectors.
+ */
````
SassDoc typically uses a triple-slash (`///`) rather than block-style comments (`/** */`) since the latter are _CSS_-comments that can be emitted into the output CSS rather than _Sass_ comments that are always omitted.

gkalpak replied:
Fixed.

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

jessicajaniuk reviewed (approved):
Reviewed-for: public-api

gkalpak committed:
fix(docs-infra): use relative URLs for internal links on error pages

Previously, some of the links on the error pages had URLs prefixed with
`https://angular.io/`. This caused them to be treated as external URLs,
which had the following downsides:
- Links would always point to `angular.io` instead of the same version
  as the error page (e.g. `next.angular.io` or `v11.angular.io`).
- Dgeni would not be able to check that the URLs are valid (i.e. point
  to existing pages).
- An external link icon would incorrectly be shown next to the links on
  `angular.io`.

This commit fixes the links to use relative URLs.

gkalpak committed:
fix(docs-infra): show external link icons for external links on error pages

This commit adds error pages to the list of docs pages that have an
external link icon next to links to external URLs.

gkalpak committed:
fix(docs-infra): show the "Errors List" page under "ERRORS" in search results

Previously, the "Errors List" page was appearing under the "OTHER"
section in search results.

This commit fixes this to make it appear under the "ERRORS" section.

Before: ![search-results before][1]
After: ![search-results after][2]

[1]: https://user-images.githubusercontent.com/8604205/107691151-c1b8ed80-6cb3-11eb-8079-fcace685f0ec.png
[2]: https://user-images.githubusercontent.com/8604205/107691145-bfef2a00-6cb3-11eb-8523-b2747fa40469.png

gkalpak committed:
fixup! fix(docs-infra): merge `docs` with `guide` and `start` with `tutorial` in search-results

gkalpak committed:
fix(docs-infra): merge `docs` with `guide` and `start` with `tutorial` in search-results

Previously, the `docs.md` guide was appearing under "OTHER" in search
results and the results for the `/start*` tutorial pages were appearing
under "START".

This commit changes this so that `docs.md` appears under "GUIDES" and
`/start*` appear under "TUTORIALS", since that is where they belong
conceptually.

It also changes the header of the guides search-area from "GUIDE" to
"GUIDES" and that of tutorials from "TUTORIAL" to "TUTORIALS".

Before: ![search-results areas before][1]
After: ![search-results areas after][2]

[1]: https://user-images.githubusercontent.com/8604205/107811568-0ce80480-6d77-11eb-8652-e7a947c36e63.png
[2]: https://user-images.githubusercontent.com/8604205/107811569-0eb1c800-6d77-11eb-9a69-0000a3703c8a.png

gkalpak committed:
refactor(docs-infra): remove unused CSS for the version selector

This commit removes some CSS rules targeting `.doc-version select` in
the sidenav. These rules do not match any elements any more, since now
we use a custom `<aio-select>` component (instead of the `<select>`
element).

gkalpak committed:
refactor(docs-infra): remove unnecessary `font-family` styles

This commit removes some unnecessary styles setting `font-family` to
`$main-font`. These styles are redundant, because the targeted elements
already inherit this style from `<body>`.

gkalpak committed:
fix(docs-infra): ensure the main font is applied to inputs

It turns out that `<input>` and `<button>` elements do not inherit the
`font-family` style from `<body>` by default, but rather use a
user-agent defined style. This means that their font-family might be
different than the one used in the rest of the page.

This commit fixes it by ensuring `<input>` (and other) elements inherit
their `font-family` style from their parent element.

Before: ![inputs font before][1]
After: ![inputs font after][2]

(The difference in font is subtle, but it's there.)

[1]: https://user-images.githubusercontent.com/8604205/107853245-76bae980-6e1d-11eb-8318-e5f6e13876cc.png
[2]: https://user-images.githubusercontent.com/8604205/107853246-77538000-6e1d-11eb-86f2-e3e7e41158f5.png

gkalpak committed:
fix(docs-infra): add spacing between past and future events sections

This commit adds some more spacing between the past and future events
sections on the "Events" page.

Before: ![events spacing before][1]
After: ![events spacing after][2]

[1]: https://user-images.githubusercontent.com/8604205/107989711-82eca580-6fdb-11eb-837e-1255d439d51a.png
[2]: https://user-images.githubusercontent.com/8604205/107989708-81bb7880-6fdb-11eb-8cf4-4653254e1c37.png

gkalpak committed:
refactor(docs-infra): create mixin for styling docs pages

Previously, in order to apply some styles to docs (i.e. non-marketing)
pages, we listed the various `.folder-*` classes that corresponded to
docs pages. This meant that adding/removing a docs area required updates
in several places, which is error-prone.

This commit avoids this by using a Sass mixin for applying styles to
docs pages.

gkalpak committed:
fix(docs-infra): make top-nav consistent across all marketing pages

Previously, some of the marketing pages had different styles for the
top-nav than others (even if they had the same layout and
blue-background header). More specifically, the top-nav had a box-shadow
and it was absolutely positioned on some marketing pages, while it had
no box-shadow and was statically positioned on others.

This commit makes the appearance of marketing pages wrt the top-nav
consistent across all marketing pages by changing the styles for the
remaining pages:
- Contributors (`/about`)
- Contribute (`/contribute`)
- Press kit (`/presskit`)

Before: ![contribute topnav shadow before][1]
After: ![contribute topnav shadow after][2]

[1]: https://user-images.githubusercontent.com/8604205/107984898-a6aaee00-6fd1-11eb-8bf3-79393c8983ff.png
[2]: https://user-images.githubusercontent.com/8604205/107984900-a7438480-6fd1-11eb-8d9b-a643d69ab692.png

gkalpak committed:
fix(docs-infra): use consistent padding in marketing pages

This commit makes the padding of the "Events", "Features" and
"Press kit" pages consistent with that of other marketing pages.

Before: ![presskit padding before][1]
After: ![presskit padding after][2]

[1]: https://user-images.githubusercontent.com/8604205/107986067-f8547800-6fd3-11eb-9164-51793e431d05.png
[2]: https://user-images.githubusercontent.com/8604205/107986064-f7234b00-6fd3-11eb-9050-ed7ec04e7443.png

gkalpak committed:
fix(docs-infra): make the "Contributors" page header similar to other marketing pages

Before (wide screen): ![contributors (wide screen) before][1]
Before (narrow screen): ![contributors (narrow screen) before][2]
After (wide screen): ![contributors (wide screen) after][3]
After (narrow screen): ![contributors (narrow screen) after][4]

[1]: https://user-images.githubusercontent.com/8604205/107983880-634f8000-6fcf-11eb-8ad9-5a5df65d3d5e.png
[2]: https://user-images.githubusercontent.com/8604205/107983893-6a768e00-6fcf-11eb-9ccd-158ec491404f.png
[3]: https://user-images.githubusercontent.com/8604205/107983903-6f3b4200-6fcf-11eb-94e1-182893b7a715.png
[4]: https://user-images.githubusercontent.com/8604205/107983895-6ba7bb00-6fcf-11eb-9ff5-59d221ba906d.png

gkalpak committed:
refactor(docs-infra): remove duplicate or unused CSS styles

This commit removes some CSS styles that have no effect (i.e. are
duplicates or overridden by other rules).

gkalpak committed:
fix(docs-infra): fix top margin of "Press kit" page on small screens

Previously, the "Press kit" page had a larger top margin on smaller
screens, that seemed unnecessary.

This commit removes the extra top margin.

Before: ![presskit top section margin before][1]
After: ![presskit top section margin after][2]

[1]: https://user-images.githubusercontent.com/8604205/107988545-efb27080-6fd8-11eb-91d6-79651f2dffaf.png
[2]: https://user-images.githubusercontent.com/8604205/107988547-f04b0700-6fd8-11eb-9c4b-d444b39a34fe.png

gkalpak committed:
refactor(docs-infra): create mixin for styling marketing pages

Previously, in order to apply some styles to marketing (i.e. non-docs)
pages, we listed the various `.page-*` classes that corresponded to docs
pages. This meant that adding/removing a marketing page required updates
in several places, which is error-prone.

This commit avoids this by using a Sass mixin for applying styles to
marketing pages.

gkalpak committed:
fixup! refactor(docs-infra): create mixin for styling docs pages

gkalpak committed:
fixup! refactor(docs-infra): create mixin for styling marketing pages

gkalpak committed:
refactor(docs-infra): convert Sass mixin from camelCase to kebab-case to follow Sass conventions

This commit converts the last remaining camelCased Sass mixin
(`deployTheme`) to kebab-case (`deploy-theme`) to follow the Sass
conventions.

Discussed in
https://github.com/angular/angular/pull/40881#discussion_r577961617.

mary-poppins commented:
You can preview 4bbe14d at https://pr40881-4bbe14d.ngbuilds.io/.
